src/main/java/com/google/collegeplanner/servlets/PlannerServlet.java

```diff
@@ -29,35 +35,209 @@
+   */
+  private void constructGraphs(JSONArray selectedClasses)
+      throws IOException, NumberFormatException {
+    // Initialize graph representation and courseList
+    indegree = new HashMap<String, Integer>();
+    corequisites = new HashMap<String, HashSet<String>>();
+    depths = new HashMap<String, Integer>();
+    credits = new HashMap<String, Integer>();
+    nextCourses = new HashMap<String, HashSet<String>>();
+    courseList = new ArrayList<>();
+
+    // creates CourseList and initialize nextCourses HashSets
+    for (Object course : selectedClasses) {
+      String key = (String) ((JSONObject) course).get("course_id");
+      courseList.add(key);
+      nextCourses.put(key, new HashSet<>());
+    }
+
+    // populate class variables
+    for (Object course : selectedClasses) {
+      String key = (String) ((JSONObject) course).get("course_id");
+      int creditVal = Integer.parseInt((String) ((JSONObject) course).get("credits"));
+      credits.put(key, creditVal);
+      totalCredits += creditVal;
+
+      HashSet<String> prereqs = getCoursesFromString(
+          (String) ((JSONObject) ((JSONObject) course).get("relationships")).get("prereqs"));
+      indegree.put(key, prereqs.size());
+
+      HashSet<String> coreqs = getCoursesFromString(
+          (String) ((JSONObject) ((JSONObject) course).get("relationships")).get("coreqs"));
+      corequisites.put(key, coreqs);
+
+      for (String prereq : prereqs) {
+        nextCourses.get(prereq).add(key);
+      }
+    }
+
+    fillDepths();
+
+    // sort courseList from greatest depth to lowest depth
+    Collections.sort(courseList,
+        (String course1, String course2) -> depths.get(course2).compareTo(depths.get(course1)));
+  }
+
+  /**
+   * Populate this.depths with the depth of each course
+   */
+  private void fillDepths() {
+    // Calls recursive function to find depths
+    for (String course : courseList) {
+      if (!depths.containsKey(course)) {
+        getDepth(course);
+      }
+    }
+  }
+
+  /**
+   * Dynamic Programming Algorithm to find the depth of a course
+   *
+   * @param course The course ID for the course
+   */
+  private int getDepth(String course) {
+    if (depths.containsKey(course)) {
+      return depths.get(course);
+    }
+    if (nextCourses.get(course).size() == 0) {
+      depths.put(course, 1);
+      return 1;
+    }
+    ArrayList<Integer> childrenDepths = new ArrayList<>();
+    for (String nextCourse : nextCourses.get(course)) {
+      childrenDepths.add(getDepth(nextCourse));
+    }
+    int courseDepth = Collections.max(childrenDepths) + 1;
+    depths.put(course, courseDepth);
+    return courseDepth;
+  }
+
+  /**
+   * Takes English representation of course prerequisites/corequisites and returns a HashSet with
+   * all the contained courses
+   *
+   * @param engCourses The natural language string to get courses from
+   */
+  private HashSet<String> getCoursesFromString(String engCourses) throws IOException {
```

Comment:
we try not to unnecessarily abbreviate unless it's super common (like http).

This should be `engCourses` -> `englishCourses`. I initially read this as `engineeringCourses`.

Or maybe it would be even clearer as `courseRelationshipsDescription` or something? Since `*Courses` implies the Course.java object that we're about to add.